common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -214,168 +107,92 @@ public void run() {
-        getActivePivSessionAsync(request, new IPivSessionCallback() {
+
+        mSmartcardCertBasedAuthManager.attemptDeviceSession(new ISmartcardCertBasedAuthManager.ISessionCallback() {
             @Override
-            public void onGetSession(@NonNull PivSession piv) {
-                try {
-                    if (piv.getPinAttempts() == 0) {
-                        Logger.info(methodTag,  "User has reached the maximum failed attempts allowed.");
-                        mDialogHolder.showErrorDialog(
-                                R.string.smartcard_max_attempt_dialog_title,
-                                R.string.smartcard_max_attempt_dialog_message);
-                        request.cancel();
-                        return;
-                    }
-                    //Create List that contains cert details only pertinent to the cert picker.
-                    final List<YubiKitCertDetails> certList = getCertDetailsFromKey(piv);
-                    //If no certs were found, cancel flow.
-                    if (certList.isEmpty()) {
-                        Logger.info(methodTag,  "No PIV certificates found on YubiKey device.");
-                        mDialogHolder.showErrorDialog(
-                                R.string.smartcard_no_cert_dialog_title,
-                                R.string.smartcard_no_cert_dialog_message);
-                        request.cancel();
-                        return;
-                    }
-                    //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
-                    mDialogHolder.showCertPickerDialog(
-                            certList,
-                            getSmartcardCertPickerDialogPositiveButtonListener(request),
-                            new SmartcardCertPickerDialog.CancelCbaCallback() {
-                                @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
-                                @Override
-                                public void onCancel() {
-                                    mDialogHolder.dismissDialog();
-                                    request.cancel();
-                                }
-                            });
-                } catch(final IOException | ApduException | BadResponseException e) {
-                    Logger.error(methodTag, e.getMessage(), e);
-                    //Show general error dialog.
+            public void onGetSession(@NonNull ISmartcardSession session) throws Exception {
+                if (session.getPinAttemptsRemaining() == 0) {
+                    indicateTooManyFailedAttempts(methodTag);
+                    request.cancel();
+                    return;
+                }
+                //Create List that contains cert details only pertinent to the cert picker.
+                final List<ICertDetails> certList = session.getCertDetailsList();
```

Comment:
Can this ever be `null`? Do we need to check?

Reply:
`getCertDetailsList()` should never return null (method has @NonNull annotation). In the case of no certs on smartcard, an empty list is returned.